File: 148. Sort List/main.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        array = list()
        node = head

        while node:
            array.append(node.val)
            node = node.next

        array.sort()
        node = head
        index = 0

        while node:
            node.val = array[index]
            index += 1
            node = node.next

        return head

File: 148. Sort List/main.rs
struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn sort_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut head = head;
        let mut current = head.as_ref();
        let mut array = Vec::new();

        while let Some(node) = current {
            array.push(node.val);
            current = node.next.as_ref();
        }

        array.sort();
        let mut current = head.as_mut();
        let mut index = 0;

        while let Some(node) = current {
            node.val = array[index];
            index += 1;
            current = node.next.as_mut();
        }

        head
    }
}
